src/cargo/core/compiler/fingerprint.rs

```diff
@@ -1,10 +1,174 @@
+//! The Metadata contains information that is required to keep Units separate.
+//! The Fingerprint includes additional information that should cause a
+//! recompile, but it is desired to reuse the same filenames. Generally the
+//! items in the Metadata do not need to be in the Fingerprint. A comparison
+//! of what is tracked:
+//!
+//! Value                                      | Fingerprint | Metadata
+//! -------------------------------------------|-------------|----------
+//! rustc                                      | ✓           | ✓
+//! Profile                                    | ✓           | ✓
+//! `cargo rustc` extra args                   | ✓           | ✓
+//! CompileMode                                | ✓           | ✓
+//! Target Name                                | ✓           | ✓
+//! Target Kind (bin/lib/etc.)                 | ✓           | ✓
+//! Enabled Features                           | ✓           | ✓
+//! Immediate dependency’s hashes              | ✓[^1]       | ✓
+//! Target or Host mode                        |             | ✓
+//! __CARGO_DEFAULT_LIB_METADATA[^4]           |             | ✓
+//! authors, description, homepage             |             | ✓[^2]
+//! package_id                                 |             | ✓
+//! Target src path                            | ✓           |
+//! Target path relative to ws                 | ✓           |
+//! Target flags (test/bench/for_host/edition) | ✓           |
+//! Edition                                    | ✓           |
+//! -C incremental=… flag                      | ✓           |
+//! mtime of sources                           | ✓[^3]       |
+//! RUSTFLAGS/RUSTDOCFLAGS                     | ✓           |
+//!
+//! [^1]: Build script and bin dependencies are not included.
+//!
+//! [^2]: This is a bug, see https://github.com/rust-lang/cargo/issues/6208
+//!
+//! [^3]: The mtime is only tracked for workspace members and path
+//!       dependencies. Git dependencies track the git revision.
+//!
+//! [^4]: `__CARGO_DEFAULT_LIB_METADATA` is set by rustbuild to embed the
+//!        release channel (bootstrap/stable/beta/nightly) in libstd.
+//!
+//! ## Fingerprint files
+//!
+//! Fingerprint information is stored in the
+//! `target/{debug,release}/.fingerprint/` directory. Each Unit is stored in a
+//! separate directory. Each Unit directory contains:
+//!
+//! - A file with a 16 hex-digit hash. This is the Fingerprint hash, used for
+//!   quick loading and comparison.
+//! - A `.json` file that contains details about the Fingerprint. This is only
+//!   used to log details about *why* a fingerprint is considered dirty.
+//!   `RUST_LOG=cargo::core::compiler::fingerprint=trace cargo build` can be
+//!   used to display this log information.
+//! - A "dep-info" file which contains a list of source filenames for the
+//!   target. This is produced by `rustc`'s `--emit=dep-info` flag. Cargo uses
```

Comment:
I think technically we actually rewrite the output of the compiler, right? IIRC we have a binary-encoding (ish) which puts the cwd at the top and then all the paths are rewritten as relative to that (or something like that)

Reply:
Correct, I have tried to clarify it.